### DISCUSSION TITLE
Default property values from database?

### DISCUSSION THREAD

**Discussion by ragulka** [spatie/laravel-data#268] [Q&A]
Is there a way to provide default property values from the database? I have a use-case where a property must be present on the data object for it to be valid, but if not provided by the user, it should default to a predefined value on user's settings, which are stored in DB.

Ie, something like this:

```php
class SongData extends Data
{
    public function __construct(
        public ?string $title,
        public ?string $artist,
        public array $tags,
    ) {
    }

    public static function defaults() : array
    {
        return [
            'tags' => auth()->user->default_song_tags, // a value fetched from DB
        ];
    }
}
```

**Comment by Nir-An** [comment]
Hi @ragulka 


you can use ``??``

```php
class SongData extends Data
{
    public function __construct(
        public string $title,
        public string $artist,
    ) {
    }
    
    public static function fromModel(Song $song): self
    {
        return new self(
          tags: $song->tags ?? auth()->user->default_song_tags,
        );
    }
}

```

**Reply by ragulka** [reply]
This only covers the use case of creating the data object from a model. The Data object will most likely be created from a request, but it may be created manually as well. My Question is - is there a way to provide default values that apply to all possible instantiation methods?

**Comment by Nir-An** [comment]
> is there a way to provide default values that apply to all possible instantiation methods?

No, as you already understand you can create scenarios for all different types. 

the default value (and the source of the value) for a request could be different from the default value for a model or custom array and so on .
personally, I prefer it this way.

Maybe change the discussion category from question to Idea? it's up to you.

**Comment by rubenvanassche** [comment]
You could always create a pipe which uses such a method and adds the values to the payload. It looks a bit overkill and opening a box of pandora of possible outcomes when we add it to the package.
